nixos/lib/utils.nix

```diff
@@ -45,6 +45,23 @@ rec {
    replaceChars ["/" "-" " "] ["-" "\\x2d" "\\x20"]
    (removePrefix "/" s);
 
+  # Quotes an argument for use in Exec* service lines.
+  # systemd accepts "-quoted strings with escape sequences, toJSON produces
+  # a subset of these.
+  # Additionally we escape % to disallow expansion of % specifiers. Any lone ;
+  # in the input will be turned it ";" and thus lose its special meaning.
+  # Every $ is escaped to $$, this makes it unnecessary to disable environment
+  # substitution for the directive.
+  escapeSystemdExecArg = arg:
+    let
+      s = builtins.toJSON (toString arg);
```

Comment:
these two are designed to mimic escapeShellArgs closely (since they have a roughly similar purpose), so we took the `toString` from there as well. `escapeShellArgs` also allows ints and floats, these ideally should as well.

how about this? then we'd copy paths when they're used, but treat the rest the same.
```suggestion
      s = builtins.toJSON (if isPath arg then "${arg}" else toString arg);
```

Reply:
That makes sense. Maybe disallow attrsets and lists though, because those must be converted to a string beforehand anyway, and it's not something we should do for the caller. `null` should probably be an error or filtered out like `builtins.derivation` does.
At this point I think we'd better explicitly enumerate the types we do support.

Reply:
then we'll stray pretty far from what escapeShellArgs does. we'd rather further your proposal of better typed representations of the Exec directives than go down that route, that seems like a more future-proof solution anyway.